src/main/java/com/coders/repository/ProductRepository.java

```diff
@@ -5,82 +5,54 @@
 import com.coders.exceptions.NoDataException;
 import com.coders.exceptions.NotSuchElementException;
 import com.coders.exceptions.TheSameNameException;
+import com.coders.helpers.ProductValidator;
 
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicInteger;
 
-import static java.lang.System.err;
-
 public final class ProductRepository {
 
     private static ProductRepository INSTANCE;
+    private ProductValidator productValidator;
     private Map<Integer, Product> productMap;
     private AtomicInteger idGenerator;
 
     private ProductRepository() {
         productMap = new HashMap<>();
         idGenerator = new AtomicInteger(1);
+        productValidator = new ProductValidator();
     }
 
+
     public static ProductRepository getInstance() {
         if (INSTANCE == null) {
             INSTANCE = new ProductRepository();
         }
-
         return INSTANCE;
     }
 
-    public Product addProduct(Product product) throws TheSameNameException, NoDataException, InvalidTypeOfDataException {
+    public Product addProduct(Product product) throws TheSameNameException, NoDataException, InvalidTypeOfDataException, NotSuchElementException {
         try {
-            if (product.getName() == null || product.getName().isEmpty()) {
-                throw new NoDataException("Product name is missing or empty.");
-            }
-            if (!(product.getName() instanceof String)) {
-                throw new InvalidTypeOfDataException("The name of the product should be a string ");
-            }
+            productValidator.validate(product);
             if (productMap.values().stream().anyMatch(prod -> prod.getName().equals(product.getName()))) {
                 throw new TheSameNameException("Product with the same name already exists: " + product.getName());
             }
-
             int productId = idGenerator.getAndIncrement();
             product.setId(productId);
             productMap.put(product.getId(), product);
-        } catch (NoDataException | TheSameNameException | InvalidTypeOfDataException e) {
-            err.println(e.getMessage());
+            return product;
+        } catch (NoDataException | NotSuchElementException | TheSameNameException | InvalidTypeOfDataException e) {
             throw e;
         }
-        return product;
     }
 
     public Product getProductById(int id) throws NoDataException, NotSuchElementException {
         try {
             Product product = productMap.get(id);
```

Comment:
tutaj najpierw wyciągasz produkt z mapy a później go walidujesz co myślę, że jest nielogiczne bo chcemy walidować przez zapisaniem do mapy więc możemy uznać, że produkt w mapie jest poprawny.
Jedyne co mamy do sprawdzenia to czy produkt istnieje w mapie o takim id.

Możemy sprawdzić if (product == null) i rzucic wyjatek jeśli tak.
Możemy tez to zrobić bardziej funkcyjnie korzystając z Optionala:

return Optional.ofNullable(productMap.get(id))
        .orElseThrow(() -> new NoDataException("No product found with ID: " + id));